clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3160,31 +3160,128 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
+            << CI << /*positive*/ 0;
+        return nullptr;
+      }
+    }
+
+    return E;
+  };
 
-    if (WGSize[i] == 0) {
-      S.Diag(AL.getLoc(), diag::err_attribute_argument_is_zero)
-          << AL << AL.getArgAsExpr(i)->getSourceRange();
+  // Check all three argument values, and if any are bad, bail out. This will
+  // convert the given expressions into constant expressions when possible.
+  XDim = CheckAndConvertArg(XDim);
+  YDim = CheckAndConvertArg(YDim);
+  ZDim = CheckAndConvertArg(ZDim);
+  if (!XDim || !YDim || !ZDim)
+    return;
+
+  // If the attribute was already applied with different arguments, then
+  // diagnose the second attribute as a duplicate and don't add it.
+  if (const auto *Existing = D->getAttr<WorkGroupSizeHintAttr>()) {
+    // The lambda returns a DupArgRes value; Same means the args have the same
+    // value, Different means the args do not have the same value, and Unknown
+    // means that the args cannot (yet) be compared.
+    enum DupArgRes { Unknown, Same, Different };
+    auto AreArgValuesIdentical = [&](const Expr *LHS, const Expr *RHS) {
+      // If either operand is still value dependent, we can't test anything.
+      const auto *LHSCE = dyn_cast<ConstantExpr>(LHS);
+      const auto *RHSCE = dyn_cast<ConstantExpr>(RHS);
+      if (!LHSCE || !RHSCE)
+        return Unknown;
+
+      // Otherwise, test that the values.
+      return LHSCE->getResultAsAPSInt() == RHSCE->getResultAsAPSInt()
+                 ? Same
+                 : Different;
+    };
+
+    DupArgRes Results[] = {AreArgValuesIdentical(XDim, Existing->getXDim()),
+                           AreArgValuesIdentical(YDim, Existing->getYDim()),
+                           AreArgValuesIdentical(ZDim, Existing->getZDim())};
+    // If any of the results are known to be different, we can diagnose at this
+    // point and drop the attribute.
+    if (llvm::is_contained(Results, Different)) {
+      Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+      Diag(Existing->getLoc(), diag::note_previous_attribute);
+      return;
+    }
+    // If all of the results are known to be the same, we can silently drop the
+    // attribute. Otherwise, we have to add the attribute and resolve its
+    // differences later.
+    if (llvm::all_of(Results, [](DupArgRes V) { return V == Same; }))
       return;
+  }
+
+  D->addAttr(::new (Context)
+                 WorkGroupSizeHintAttr(Context, CI, XDim, YDim, ZDim));
+}
+
+WorkGroupSizeHintAttr *
+Sema::MergeWorkGroupSizeHintAttr(Decl *D, const WorkGroupSizeHintAttr &A) {
+  // Check to see if there's a duplicate attribute already applied.
+  if (const auto *DeclAttr = D->getAttr<WorkGroupSizeHintAttr>()) {
+    // The lambda returns a DupArgRes value; Same means the args have the same
+    // value, Different means the args do not have the same value, and Unknown
+    // means that the args cannot (yet) be compared.
+    enum DupArgRes { Unknown, Same, Different };
```

Comment:
Looks like there is quite a bit of code duplication. Does it make sense to move all this to a function?

Reply:
I had gone back and forth with that and went with this approach, but I think I can split the lambda out pretty easily. The issue is that a common function would struggle to issue the diagnostics because one call takes an semantic attribute pointer and the other takes a parsed attribute reference, so the call to `Diag()` gets pretty awkward.

I made it a bit better, WDYT?

Reply:
I think this is good. Thanks for making the change!